org.es_de.emulationstation-de.yml

```diff
@@ -0,0 +1,82 @@
+#  SPDX-License-Identifier: MIT
+#
+#  EmulationStation Desktop Edition
+#  org.es_de.emulationstation-de.yml
+#
+#  Flatpak manifest file for use with flatpak-builder.
+#
+
+app-id: org.es_de.emulationstation-de
+sdk: org.freedesktop.Sdk
+runtime: org.freedesktop.Platform
+runtime-version: '21.08'
+command: emulationstation
+branch: stable
+finish-args:
+  - --share=ipc
+  - --socket=x11
+  - --socket=wayland
+  - --socket=pulseaudio
+  - --device=all
+  - --filesystem=host:ro
+  - --filesystem=/var/lib/flatpak:ro
+  - --filesystem=home
+  - --filesystem=~/.var/app
```

Comment:
or even why do you need this one?

Reply:
This is required to locate RetroArch cores as they are installed in ~/.var/app

Reply:
so basically you want to grant access to all apps for one.

Reply:
Yes, without this access my application will not be able to launch all the emulators that are installed as Flatpaks, which nowadays is the preferred way to install emulators. Especially so on the Steam Deck where many users run my application today (it's currently distributed as an AppImage).

Reply:
1. it's included in the broad `--filesystem=home`
2. I think you misunderstand how flatpak works. It's a sandbox. If you want to launch another flatpak it's done outside of that sandbox.

Reply:
Yes that is a bit strange as ~/.var/app is a subdirectory in home, but if I remove this entry then it doesn't work. I can relocate the RetroArch cores to any location outside of ~/.var/app, for instance ~/.var/cores but as soon as they are placed inside and subdirectory in ~/.var/app they can no longer be found. Unless I add this entry, then it works.

No I think I understand how it works, I launch emulators using "flatpak-spawn --host" so they run outside the sandbox, but I need to be able to check which cores are available before triggering the launch of the emulator as I warn the user with a popup notification about missing cores. So therefore I need to be able to see the files inside this directory.

Reply:
This does seem excessive though, and I'm not sure if I understand the point. If this is about cores downloaded by ES itself, it should really use $XDG_DATA_HOME to store them.

Reply:
No I don't download any cores, my application only acts as a frontend. So what I'll do is that on game launch check if the core that is configured for the certain game/system exists, and if it doesn't I display an error to the user. If it does exist, then I trigger a launch of the emulator using flatpak-spawn --host and pass the core location as an argument. If you have a suggestion to solve this without using - --filesystem=~/.var/app then I'm very happy to accommodate that, but it has to be a solution that still makes the application work correctly.

Reply:
Btw, the reason I don't set explicit links to the RetroArch config directory is that my application is very configurable and the user may create their own custom systems configuration entries. So they could use another application/emulator potentially that works with emulator cores or equivalent functionality, and just pointing to a specific RetroArch directory would cripple the ability to customize systems. As well, my frontend is not limited to games and could launch any application installed by the users, even desktop applications.

Reply:
Bear with me because I clearly don't know enough about ES. By these comments, do you mean that ES will try to re-use RetroArch cores that were downloaded with flatpaked RetroArch and saved within ~/.var/app?

Reply:
OK, I think I understand that correctly based on https://gitlab.com/leonstyhre/emulationstation-de/-/blob/master/resources/systems/unix/es_find_rules.xml#L23. This could really be limited to `~/.var/app/org.libretro.RetroArch` then.

Reply:
That being said, I placed a GBA rom in `~/ROMs/gba`, and it shows up on the list, but fails to launch, saying it couldn't find an emulator.

```
Apr 28 10:23:59 Error:	Couldn't launch game, emulator binary not found
Apr 28 10:23:59 Error:	Raw emulator launch command:
Apr 28 10:23:59 Error:	%EMULATOR_RETROARCH% -L %CORE_RETROARCH%/mgba_libretro.so %ROM%
```

The XML file defines an entry for mGBA though, so I think I understand even less what these definitions are about.

Reply:
…and if it defaults to RetroArch, why not bundle it with the app?

Reply:
Installing RetroArch as Flatpak is only one possibility, the user may have installed it as a Snap, AppImage or via the OS repository, or possibly built it from source code. As well the cores could be relocated to wherever, possibly to an SD card, external drive, mounted NFS share or whatever. I've seen users doing all kinds of stuff by now and ES-DE is fully configurable and if customizing your es_find_rules.xml file almost anything is possible.

Many users apparently don't like RetroArch at all and only want to use ES-DE with standalone emulators, and as well I don't plan to release updates that frequently, so it would be a bad idea to ship RetroArch in the same Flatpak. I would basically have to do the same for all supported emulators. There is actually a separate project to do precisely that which is called RetroDECK, their plan is to bundle all supported emulators in the same Flatpak with ES-DE. I collaborate with this project and I think they also have an open PR here.

Reply:
When you launch a game ES-DE checks if the emulator is installed, and if not it shows an error notification inside the application. It tries to locate emulators via all the rules defined in es_find_rules.xml in sequential order. Only after the emulator binary is found it attempts to locate the cores configured using the %CORE_% variable. So in your case the emulator binary is not found for some reason, and it would be quite interesting to understand why. Starting ES-DE with the --debug flag should provide some more detailed log output. I guess you tried it with the lastest Flatpak I had built yesterday?

I had to make quite an ugly hack to get around the Flatpak sandbox limitations or I would not be able to find emulators installed via the repository which could go into directories shadowed by the sandbox, like /usr/bin or /usr/local/bin for instance. As such I call the "which" command using "flatpak-spawn --host" and it actually works although being very far from elegant. If you can let me know a better way to do it I'm very happy to implement that instead, but it really has to be able to locate binaries anywhere in the host. I have actually not implemented the same hack for absolute paths so at the moment the Flatpak build is a bit crippled as you couldn't for instance make an entry like /usr/games/chocolate-doom and have ES-DE launch that when running as a Flatpak. Either I'll add support for that to my ugly hack, or there may be a more elegant solution to get around the shadowed directories. If you have any ideas how to, then please advice!

Reply:
In general, sure ES-DE could be locked down further in the manifest, but the more restrictions are added, the less flexible the software becomes and at a certain point it's not worth shipping as a Flatpak. This is so as the AppImage, DEB, RPM or AUR releases have no such limitations. I would really like to ship as a Flatpak, but I can't cripple the application too much as I would get a large amount of support requests from users. Nobody should for example have to use Flatseal to get ES-DE running properly even if they have their ROMs on an SD card or their RetroArch cores in a non-standard location.